### PR TITLE
[go_router_builder] Add ShellRoute support to go_router_builder

### PR DESCRIPTION
**Pull request by GP4cK**
Continuing the work of https://github.com/flutter/packages/pull/3269
Fixes https://github.com/flutter/flutter/issues/111909

## Pre-launch Checklist

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [x] I read and followed the [relevant style guides] and ran the auto-formatter. (Unlike the flutter/flutter repo, the flutter/packages repo does use `dart format`.)
- [x] I signed the [CLA].
- [x] The title of the PR starts with the name of the package surrounded by square brackets, e.g. `[shared_preferences]`
- [x] I listed at least one issue that this PR fixes in the description above.
- [x] I updated `pubspec.yaml` with an appropriate new version according to the [pub versioning philosophy], or this PR is [exempt from version changes].
- [x] I updated `CHANGELOG.md` to add a description of the change, [following repository CHANGELOG style].
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] I added new tests to check the change I am making, or this PR is [test-exempt].
- [x] All existing and new tests are passing.

If you need help, consider asking for advice on the #hackers-new channel on [Discord].

[Contributor Guide]: https://github.com/flutter/packages/blob/main/CONTRIBUTING.md
[Tree Hygiene]: https://github.com/flutter/flutter/wiki/Tree-hygiene
[relevant style guides]: https://github.com/flutter/packages/blob/main/CONTRIBUTING.md#style
[CLA]: https://cla.developers.google.com/
[flutter/tests]: https://github.com/flutter/tests
[breaking change policy]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes
[Discord]: https://github.com/flutter/flutter/wiki/Chat
[pub versioning philosophy]: https://dart.dev/tools/pub/versioning
[exempt from version changes]: https://github.com/flutter/flutter/wiki/Contributing-to-Plugins-and-Packages#version-and-changelog-updates
[following repository CHANGELOG style]: https://github.com/flutter/flutter/wiki/Contributing-to-Plugins-and-Packages#changelog-style
[test-exempt]: https://github.com/flutter/flutter/wiki/Tree-hygiene#tests

### CODE REVIEW

**Comment by GP4cK** [comment]
[As mentioned in the other PR](https://github.com/flutter/packages/pull/3269#issuecomment-1463191489), please let me know how you want me to refactor the `GoRouterShellGenerator` to avoid duplicated code and I'll write a test based on it.
Cheers.

File: packages/go_router_builder/example/lib/main.g.dart [hunk 1]
```diff
@@ -52,7 +58,11 @@ extension $HomeRouteExtension on HomeRoute {
 }
 
 extension $FamilyRouteExtension on FamilyRoute {
-  static FamilyRoute _fromState(GoRouterState state) => FamilyRoute(
+  static FamilyRoute _fromState(GoRouterState state) => const FamilyRoute();
```
Comment (chunhtai) [inline packages/go_router_builder/example/lib/main.g.dart:61]: This feel a bit unnecessary? maybe we don't need to generate extension for shellRoute at all?

File: packages/go_router_builder/example/lib/main.dart [hunk 2]
```diff
@@ -35,6 +35,7 @@ class App extends StatelessWidget {
   late final GoRouter _router = GoRouter(
```
Comment (chunhtai) [inline packages/go_router_builder/example/lib/main.dart:35]: Since you have already create the another shellroute example let's revert the change to the main.dart. As a main example, I think we should not make it more complex whenever we add new feature

**Review by chunhtai — COMMENTED** [review]
Yes, it will be better if we can refactor GoRouterGenerator to prevent code duplication. I imagine we will add more shellroute subclass in the future

File: packages/go_router_builder/example/lib/main.dart (hunk 2 above)
Comment (GP4cK) [reply packages/go_router_builder/example/lib/main.dart]: I reverted the changes here: [548882a](https://github.com/flutter/packages/pull/3439/commits/548882ac28c7196aa286f05384bec1c4c34edfcf)

File: packages/go_router_builder/example/lib/main.g.dart (hunk 1 above)
Comment (GP4cK) [reply packages/go_router_builder/example/lib/main.g.dart]: This change got reverted with the changes on example/lib/main.dart but if you're talking more generally about the `ShellRouteExtension._fromState`, it is passed as the factory of `ShellRouteData.$route` to create `$appRoutes`.

**Comment by GP4cK** [comment]
> Yes, it will be better if we can refactor GoRouterGenerator to prevent code duplication. I imagine we will add more shellroute subclass in the future

I've refactored the generator in [76b9b81](https://github.com/flutter/packages/pull/3439/commits/76b9b81ec95f3a85739a30b5c797edee66a593bf), let me know what you think.

For the tests, I'm not sure what is the most pertinent. It seems most of the code is already covered by `packages/go_router_builder/test/builder_test.dart` or `packages/go_router_builder/example/test/ensure_build_test.dart`. Let me know what use case you want to test.
Thanks!

**Review by dorklein — COMMENTED** [review]
The changes look good

**Review by alanchan-dev — COMMENTED** [review]
Everything looks good.

File: packages/go_router_builder/example/lib/shell_route_example.dart [hunk 3]
```diff
@@ -0,0 +1,134 @@
+// Copyright 2013 The Flutter Authors. All rights reserved.
```
Comment (chunhtai) [inline packages/go_router_builder/example/lib/shell_route_example.dart:1]: Can you add a simple test for this example? Just make sure things will compile and the shellRoute's widget is built

File: packages/go_router_builder/lib/src/route_config.dart [hunk 4]
```diff
@@ -66,19 +68,26 @@ class RouteConfig {
...
-      throw InvalidGenerationSourceError(
-        'Missing `path` value on annotation.',
-        element: element,
-      );
-    }
+    final InterfaceType type = reader.objectValue.type! as InterfaceType;
+    // Ignore the deprected `element2` so that the "downgraded_analyze" CI step
+    // passes.
+    //ignore: deprecated_member_use
+    final bool isShellRoute = type.element2.name == 'TypedShellRoute';
```
Comment (chunhtai) [inline packages/go_router_builder/lib/src/route_config.dart:75]: this should use `element`

File: packages/go_router_builder/example/lib/shell_route_example.dart [hunk 5]
```diff
@@ -0,0 +1,134 @@
...
+class HomeScreen extends StatelessWidget {
+  const HomeScreen({super.key});
+
+  @override
+  Widget build(BuildContext context) => Scaffold(
+        appBar: AppBar(title: const Text('foo')),
+      );
+}
+
+@TypedShellRoute<MyShellRouteData>(
```
Comment (chunhtai) [inline packages/go_router_builder/example/lib/shell_route_example.dart:37]: I may have missed something, how would one specified navigator key for typedshellRoute and also the parent navigator key for typedGoRoute

File: packages/go_router_builder/lib/src/route_config.dart [hunk 6]
```diff
@@ -105,6 +120,39 @@ class RouteConfig {
   final String _path;
   final InterfaceElement _routeDataClass;
   final RouteConfig? _parent;
+  final String? _key;
+  final bool _isShellRoute;
+
+  static String? _decodeKey(InterfaceElement classElement) {
```
Comment (chunhtai) [inline packages/go_router_builder/lib/src/route_config.dart:126]: What is this function trying to grab?

File: packages/go_router_builder/lib/src/route_config.dart (hunk 4 above)
Comment (GP4cK) [reply packages/go_router_builder/lib/src/route_config.dart]: Changed to element in [4600f37](https://github.com/flutter/packages/pull/3439/commits/4600f3795b58653eff4618e96fdf0c8fbafcdb31)

File: packages/go_router_builder/example/lib/shell_route_example.dart (hunk 3 above)
Comment (GP4cK) [reply packages/go_router_builder/example/lib/shell_route_example.dart:1]: Added test in [fb449ff](https://github.com/flutter/packages/pull/3439/commits/fb449ff275a80230fc6656298eca347e744f82e6)

File: packages/go_router_builder/lib/src/route_config.dart [hunk 7]
```diff
@@ -69,10 +69,10 @@ class RouteConfig {
   ) {
     assert(!reader.isNull, 'reader should not be null');
     final InterfaceType type = reader.objectValue.type! as InterfaceType;
-    // Ignore the deprected `element2` so that the "downgraded_analyze" CI step
+    // Ignore the deprecated `element2` so that the "downgraded_analyze" CI step
```
Comment (GP4cK) [inline packages/go_router_builder/lib/src/route_config.dart]: I removed the comment in [f0924ed](https://github.com/flutter/packages/pull/3439/commits/f0924edc8da7090855bcd33a646c3e91f5d2bad6) since it no longer applies.

File: packages/go_router_builder/lib/src/route_config.dart (hunk 4 above)
Comment (GP4cK) [reply packages/go_router_builder/lib/src/route_config.dart]: Actually I think that was intended. If I use element, the `downgraded_analyze` step of the CI fails (that's the comment above). How shall I handle this?

File: packages/go_router_builder/lib/src/route_config.dart (hunk 6 above)
Comment (GP4cK) [reply packages/go_router_builder/lib/src/route_config.dart:123]: This function is grabbing the static getter $navigatorKey. Example:

```dart
@TypedShellRoute<MyShellRouteData>(
  routes: <TypedRoute<RouteData>>[ /* ... */ ],
)
class MyShellRouteData extends ShellRouteData {
  const MyShellRouteData();

  static GlobalKey<NavigatorState>? get $navigatorKey => shellKey; // <- Grabs this.

  @override
  Widget builder(BuildContext context, GoRouterState state, Widget navigator) {
    return MyShellRouteScreen(child: navigator);
  }
}
```

File: packages/go_router_builder/example/lib/shell_route_example.dart (hunk 5 above)
Comment (GP4cK) [reply packages/go_router_builder/example/lib/shell_route_example.dart:37]: I think what @johnpryan had in mind was to add a static final or getter called `$navigatorKey` to the class extending `ShellRouteData` and `GoRouteData` and this will be grabbed by the `_decodeKey` and added as the navigatorKey / parentNavigatorKey (BTW I fixed a small bug here: [6eab717](https://github.com/flutter/packages/pull/3439/commits/6eab71755acacc3e13c73902cced93ca780d9aa1))

File: packages/go_router_builder/lib/src/route_config.dart [hunk 8]
```diff
@@ -325,8 +325,11 @@ RouteBase get $_routeGetterName => ${_routeDefinition()};
         : '''
 routes: [${_children.map((RouteConfig e) => '${e._routeDefinition()},').join()}],
 ''';
-    final String navigatorKey =
-        _key == null || _key!.isEmpty ? '' : 'navigatorKey: $_key,';
+    final String navigatorKeyParameterName =
+        _isShellRoute ? 'navigatorKey' : 'parentNavigatorKey';
+    final String navigatorKey = _key == null || _key!.isEmpty
+        ? ''
+        : '$navigatorKeyParameterName: $_key,';
```
Comment (GP4cK) [inline packages/go_router_builder/lib/src/route_config.dart:332]: In order to test this, I was thinking to do the same as in `packages/go_router_builder/test/builder_test.dart` and use `@ShouldGenerate` to make sure we use the correct param name like so:

<details>
<summary>go_router_builder/test/test_inputs/_shell_test_input.dart</summary>

```dart
import 'package:flutter/src/widgets/framework.dart';
import 'package:flutter/src/widgets/navigator.dart';
import 'package:go_router/go_router.dart';
import 'package:source_gen_test/source_gen_test.dart';

@ShouldGenerate(r'''
RouteBase get $shellRouteWithKey => ShellRouteData.$route(
      factory: $ShellRouteWithKeyExtension._fromState,
      navigatorKey: ShellRouteWithKey.$navigatorKey,
      routes: [
        GoRouteData.$route(
          path: '/foo',
          factory: $GoRouteWithKeyExtension._fromState,
          parentNavigatorKey: GoRouteWithKey.$navigatorKey,
        ),
      ],
    );

extension $ShellRouteWithKeyExtension on ShellRouteWithKey {
  static ShellRouteWithKey _fromState(GoRouterState state) =>
      const ShellRouteWithKey();
}

extension $GoRouteWithKeyExtension on GoRouteWithKey {
  static GoRouteWithKey _fromState(GoRouterState state) =>
      const GoRouteWithKey();

  String get location => GoRouteData.$location(
        '/foo',
      );

  void go(BuildContext context) => context.go(location);

  void push(BuildContext context) => context.push(location);

  void pushReplacement(BuildContext context) =>
      context.pushReplacement(location);
}
''')
@TypedShellRoute<ShellRouteWithKey>(
  routes: <TypedRoute<RouteData>>[
    TypedGoRoute<GoRouteWithKey>(path: '/foo'),
  ],
)
class ShellRouteWithKey extends ShellRouteData {
  const ShellRouteWithKey();

  static final GlobalKey<NavigatorState> $navigatorKey =
      GlobalKey<NavigatorState>();
}

class GoRouteWithKey extends GoRouteData {
  const GoRouteWithKey();

  static GlobalKey<NavigatorState> get $navigatorKey => GlobalKey();
}
```
</details>

However, if I do this, I have a warning on the first 2 imports that `flutter` is not a dependency.
What do you prefer?
1. Ignore the warning
2. I add `flutter` to devDependencies
3. Else?

File: packages/go_router_builder/example/lib/shell_route_example.dart (hunk 5 above)
Comment (chunhtai) [reply packages/go_router_builder/example/lib/shell_route_example.dart:37]: >  static final or getter

It seems a bit not obvious. Why not just let it be a parameter into the 
`@TypedShellRoute<MyShellRouteData>` ?

File: packages/go_router_builder/lib/src/route_config.dart (hunk 4 above)
Comment (chunhtai) [reply packages/go_router_builder/lib/src/route_config.dart]: we have been using the element in route_config.dart. I think you need to add a space inbetween ignore and slash?
```
// ignore: deprecated_member_use
```

Comment (GP4cK) [reply packages/go_router_builder/lib/src/route_config.dart]: Ok I've added the same comment as in the other occurences of `element` in the code [9f8b067](https://github.com/flutter/packages/pull/3439/commits/9f8b0670b911abe83ad8bc00a13ae1ebab0bea2b)

File: packages/go_router_builder/example/lib/shell_route_example.dart (hunk 5 above)
Comment (GP4cK) [reply packages/go_router_builder/example/lib/shell_route_example.dart:37]: I agree it's not obvious. I assumed that you two had agreed on the approach beforehand and I just picked up the work where it was left off.
So if we go ahead with adding:
- `navigatorKey` to `TypedShellRoute`
- `parentNavigatorKey` to `TypedGoRoute`

I will first need to make a PR on go_router (since that's where these classes are defined) and then raise the minimum version of go_router in go_router_builder.

Could you confirm that it's ok?

**Comment by kuyacai** [comment]
hi @GP4cK , Thank you very much for your solution. There may be a little problem.
If I add another `route`, the generated file has two `List<RouteBase> get $appRoutes => [...`
For example, I add `LoginRoute` to  the bottom of  `shell_route_example.dart`:
```dart
///ignore above
/// Login route
@TypedGoRoute<LoginRoute>(path: '/login')

class LoginRoute extends GoRouteData {
  const LoginRoute();

  @override
  Widget build(BuildContext context, GoRouterState state) =>
      LoginScreen();
}
class LoginScreen extends StatelessWidget {
  const LoginScreen({super.key});

  @override
  Widget build(BuildContext context) {
    return const Text('Login');
  }
}
```
The `shell_route_example.g.dart` has two `$appRoutes`:
```dart
/// ignore other code
List<RouteBase> get $appRoutes => [
      $myShellRouteData,
    ];

/// ignore

List<RouteBase> get $appRoutes => [
      $loginRoute,
    ];
```
The correct `$appRoutes` should be:

```dart
List<RouteBase> get $appRoutes => [
      $myShellRouteData,
      $loginRoute,
    ];
```

The reason may be hear:

There are two `GoRouterGenerator` in `go_router_builder.dart`。

```dart
Builder goRouterBuilder(BuilderOptions options) => SharedPartBuilder(
      const <Generator>[
        GoRouterGenerator(
          annotation: 'TypedGoRoute',
          routeClass: 'GoRouteData',
        ),
        GoRouterGenerator(
          annotation: 'TypedShellRoute',
          routeClass: 'ShellRouteData',
        ),
      ],
      'go_router',
    );
```

but the method `generate` of `GoRouterGenerator` returns:

```dart

@override
  FutureOr<String> generate(LibraryReader library, BuildStep buildStep) async {
    
    /// ignore 

    return <String>[
      '''
List<RouteBase> get \$appRoutes => [
${getters.map((String e) => "$e,").join('\n')}
    ];
''',
      ...values,
    ].join('\n\n');
  }

```

File: packages/go_router_builder/example/lib/shell_route_example.dart (hunk 5 above)
Comment (chunhtai) [reply packages/go_router_builder/example/lib/shell_route_example.dart:37]: yes, that is ok. Sorry about the miscommunication, the original PR was still in its initial state, so there may be some issue that hasn't been discovered yet.

Comment (GP4cK) [reply packages/go_router_builder/example/lib/shell_route_example.dart:37]: No problem, I also should have double checked.
Here is the PR on go_router for the first step: https://github.com/flutter/packages/pull/3471

**Comment by GP4cK** [comment]
@kuyacai thanks a lot for this.
I've refactored the generator so that it generates only a single `$appRoutes` in [7f4c6e0](https://github.com/flutter/packages/pull/3439/commits/7f4c6e0963bb6a89184193a533e33fc1c1023302)

**Comment by kuyacai** [comment]
> 

ok, thx!

File: packages/go_router_builder/example/lib/shell_route_example.dart (hunk 5 above)
Comment (GP4cK) [reply packages/go_router_builder/example/lib/shell_route_example.dart:37]: Hey @chunhtai, actually passing the keys as parameters of `TypedShellRoute` / `TypedGoRoute` doesn't work because the keys need to be `const` and `GlobalKey<NavigatorState>()` is not...
So is it ok to go ahead with this PR as it is now and I'll make another one to revert https://github.com/flutter/packages/pull/3471?

**Comment by azimuthdeveloper** [comment]
Sorry if this PR is not the best place to comment on this, but I've been trying to use the `@TypedShellRoute` in my code, which appears in `go_router`, but seems to be ignored by `go_router_builder`. It almost seems like the bits to make this work are part of the `go_router` package, but the `go_router_builder` package has not been updated yet to actually generate them.

Is that a fair assessment of where this is at, at the moment?

**Comment by GP4cK** [comment]
> Is that a fair assessment of where this is at, at the moment?

That's correct. Right now I'm waiting for @chunhtai or @johnpryan 's advice on how to manage the `navigatorKey` and the `parentNavigatorKey`.

We cannot pass them to `TypedShellRoute` / `TypedGoRoute` because `GlobalKey<NavigatorState>()` is not a const.
So either we can either:
1. declare them as a static member called `$navigatorKey` in the typed route and the generator will pick them up (but it's not obvious for users of this library)
2. use the workaround described in https://github.com/flutter/flutter/issues/119294#issuecomment-1474867983 that clones each routes after the code generation to add the keys before passing them to GoRouter.

**Comment by azimuthdeveloper** [comment]
Ah okay. I hope it gets sorted soon because it's quite confusing at the moment to have `TypedShellRoute` available in `go_router`, but for it to have no effect when `build_runner` runs. Thanks for the update @GP4cK

**Comment by chunhtai** [comment]
Hi @GP4cK can you help me understand why it needs to be const? I couldn't find the place the enforce it

**Comment by chunhtai** [comment]
I just learned from @johnpryan that annotation has to be const. Let me think a bit on what else we can do here...

**Comment by johnpryan** [comment]
Yes, annotations must have a `const` constructor, so all of the fields must be final and all of the parameters must be const too. Since `Key`s in Flutter aren't const, I'm not sure how we're going to support `navigatorKey` or `parentNavigatorKey`.

File: packages/go_router_builder/example/lib/shell_route_example.dart (hunk 5 above)
Comment (GP4cK) [reply packages/go_router_builder/example/lib/shell_route_example.dart:37]: image

Comment (GP4cK) [reply packages/go_router_builder/example/lib/shell_route_example.dart:37]: And https://dart.dev/language/metadata

**Comment by chunhtai** [comment]
@GP4cK 

After some thought, I think we can have a getter in the GoRouteData and ShellRouteData

for example:

```
abstract class GoRouteData extends RouteData {
 ...
 GlobalKey<NavigatorKey> get $parentNavigatorKey => null;
 ...
}

abstract class ShellRouteData extends RouteData {
 ...
 GlobalKey<NavigatorKey> get $navigatorKey => null;
 ...
}
```

and user can override and provide their navigator key in the subclass if they want. I was hoping this would be more clear than using the static field. At least people can see the getter in the abstract class and notice such thing exists.

WDYT @johnpryan @GP4cK

**Comment by GP4cK** [comment]
I'm good with that and it's not too far from the initial implementation.

**Comment by kaankoken** [comment]
I have a question regarding `@TypedShellRoute`. Let's say I have two screens:  `SplashView` & `HomeView`. The first screen shown is `SplashView`, generated by `@TypedGoRouter`, after the native splash screen. After `SplashView`, it will directly transition to `HomeView` which is generated by `@TypedShellGoRouter`. 

Apparently, to run this flow, I should put `HomeView` as a `childRoute` element. If I do not, it does not generate the code.
<details>
     <summary>Code</summary>

```dart

final GlobalKey<NavigatorState> _navigatorKey = GlobalKey<NavigatorState>(); 
final GlobalKey<NavigatorState> _shellNavigatorKey = GlobalKey<NavigatorState>();

final router = GoRouter(
  navigatorKey: _navigatorKey,
  initialLocation: AppRoutes.SPLASH,
  observers: [NavigatorObservers()],
  routes: $appRoutes,
);

/// Splash Route

@immutable
@TypedGoRoute<SplashViewRoute>(path: AppRoutes.SPLASH, routes: [])
class SplashViewRoute extends GoRouteData {
  const SplashViewRoute();

  @override
  Widget build(BuildContext _, GoRouterState state) => const SplashView();
}

///

/// Home Route
@TypedShellRoute(
  routes: [
    TypedGoRoute<SubViewRoute>(path: AppRoutes.HOME),
  ],
)
class HomeViewRoute extends ShellRouteData {
  const HomeViewRoute();

  @override
  Widget builder(BuildContext context, GoRouterState state, Widget navigator) => const SizedBox.shrink();

  @override
  GlobalKey<NavigatorState>? get navigatorKey => _shellNavigatorKey;
}

class SubViewRoute extends GoRouteData {
  const SubViewRoute();

  @override
  Widget build(BuildContext _, GoRouterState state) => const SizedBox.shrink();

  @override
  GlobalKey<NavigatorState>? get navigatorKey => _shellNavigatorKey;
}

```
</details>

If I turned `HomeView` to `@TypedGoRoute`, I can navigate without declaring it as a `childRoute`.
Am I doing something wrong? Should not I at least navigate to `SubView`?

**Comment by chunhtai** [comment]
yeah I was confused myself how `GoRouteData.$route` was called. it looks like using static field may be cleanest way to approach this. Sorry for all the detours, let's go with the original proposal.

File: packages/go_router_builder/lib/src/route_config.dart [hunk 9]
```diff
@@ -105,6 +120,39 @@ class RouteConfig {
   final String _path;
   final InterfaceElement _routeDataClass;
   final RouteConfig? _parent;
+  final String? _key;
+  final bool _isShellRoute;
+
+  static String? _decodeKey(InterfaceElement classElement) {
+    bool whereStatic(FieldElement element) => element.isStatic;
+    bool whereKeyName(FieldElement element) => element.name == r'$navigatorKey';
```
Comment (chunhtai) [inline packages/go_router_builder/lib/src/route_config.dart:128]: I would expect GoRouteData to use $parentNavigatorKey and  ShellRouteData to use $navigatorKey to match the API of GoRoute and ShellRoute

File: packages/go_router_builder/lib/src/route_config.dart (hunk 6 above)
Comment (chunhtai) [inline packages/go_router_builder/lib/src/route_config.dart:126]:
```suggestion
  static String? _generateNavigatorKeyGetterCode(InterfaceElement classElement) {
```

File: packages/go_router_builder/example/lib/shell_route_example.dart [hunk 10]
```diff
@@ -0,0 +1,153 @@
...
+  routes: <TypedRoute<RouteData>>[
+    TypedGoRoute<FooRouteData>(path: '/foo'),
+    TypedGoRoute<BarRouteData>(path: '/bar'),
+  ],
+)
+class MyShellRouteData extends ShellRouteData {
```
Comment (chunhtai) [inline packages/go_router_builder/example/lib/shell_route_example.dart:43]: Can you also add navigatorKey and parentNavigatorKey example? it can be in this example or a separate one.

File: packages/go_router_builder/lib/src/route_config.dart (hunk 8 above)
Comment (chunhtai) [reply packages/go_router_builder/lib/src/route_config.dart:335]: Just add an example in the example folder, and add a test for that example to make sure everything works correctly. We also have a test to ensure the .g.dart is always the latest code. So if someone breaks the route_config.dart either the example test will fail or the .g.dart test will fail.

**Review by chunhtai — COMMENTED** [review]
Thanks for the work, I left some comments, but overall looks great.

Can you add a section in readme to talk about shellroute and navigatorkey?

File: packages/go_router_builder/lib/src/route_config.dart (hunk 9 above)
Comment (GP4cK) [reply packages/go_router_builder/lib/src/route_config.dart]: Done in [61c3d9d](https://github.com/flutter/packages/pull/3439/commits/61c3d9d077012650e0ce82ffb51f25c0892bac20)

File: packages/go_router_builder/lib/src/route_config.dart (hunk 6 above)
Comment (GP4cK) [reply packages/go_router_builder/lib/src/route_config.dart]: Done in https://github.com/flutter/packages/pull/3439/commits/61c3d9d077012650e0ce82ffb51f25c0892bac20

File: packages/go_router_builder/example/lib/shell_route_example.dart (hunk 10 above)
Comment (GP4cK) [reply packages/go_router_builder/example/lib/shell_route_example.dart:43]: Done in [Add example of navigation with keys](https://github.com/flutter/packages/pull/3439/commits/3013a05f73a178e8926b3a7f3f4e7158a373ac0b).

Here's a quick demo of the example. I use the navigator key to make sure that the dialog route is displayed over the entire screen instead of just the right of the navigator rail:
nav with keys

I've also added a section in the README in [Add section about navigator keys in README](https://github.com/flutter/packages/pull/3439/commits/3376c7367804892dbe9cc1745ff2966778107f94)

**Review by johnpryan — APPROVED** [review]

File: packages/go_router_builder/example/pubspec.yaml [hunk 11]
```diff
@@ -12,7 +12,7 @@ dependencies:
   provider: 6.0.5
 
 dev_dependencies:
-  build_runner: ^2.0.0
```
Comment (chunhtai) [inline packages/go_router_builder/example/pubspec.yaml:15]: Is there a reason this need to be a 2.3.0?

File: packages/go_router_builder/example/test/all_types_test.dart [hunk 12]
```diff
@@ -2,7 +2,6 @@
 // Use of this source code is governed by a BSD-style license that can be
 // found in the LICENSE file.
 
-// import 'package:flutter/material.dart';
```
Comment (chunhtai) [inline packages/go_router_builder/example/test/all_types_test.dart:5]: Thanks!

File: packages/go_router_builder/example/lib/shell_route_with_keys_example.dart [hunk 13]
```diff
@@ -0,0 +1,169 @@
+// Copyright 2013 The Flutter Authors. All rights reserved.
```
Comment (chunhtai) [inline packages/go_router_builder/example/lib/shell_route_with_keys_example.dart:1]: Can you also add a test for this? just to make sure the code can actually compile

File: packages/go_router_builder/README.md [hunk 14]
```diff
@@ -327,3 +327,42 @@ class FancyRoute extends GoRouteData {
...
+class MyGoRouteData extends GoRouteData {
+  const MyGoRouteData();
+
+  static final GlobalKey<NavigatorState> $parentNavigatorKey = rootNavigatorKey;
+
+  @override
+  Widget build(BuildContext context, GoRouterState state) {
+    // ...
+  }
+}
```
Comment (chunhtai) [inline packages/go_router_builder/README.md:367]: Maybe also add a link to the examples

File: packages/go_router_builder/example/pubspec.yaml (hunk 11 above)
Comment (GP4cK) [reply packages/go_router_builder/example/pubspec.yaml:15]: If I change it to `build_runner: 2.0.0` and run `flutter clean && flutter pub get`, I get this error:

```
Because every version of go_router_builder from path depends on analyzer >=4.4.0 <6.0.0 and build_runner >=2.0.0 <2.0.6 depends on analyzer ^1.4.0, go_router_builder from path is incompatible
  with build_runner >=2.0.0 <2.0.6.
```

Similar issue with `2.1.0`. However `2.2.0` works fine. Shall I change to `^2.2.0`?

File: packages/go_router_builder/README.md (hunk 14 above)
Comment (GP4cK) [reply packages/go_router_builder/README.md:367]: I've added the link here: [Add link to example](https://github.com/flutter/packages/pull/3439/commits/d7752fd6d2151d60b8391b7bdc3381d6ea806810)
It will work when the PR is merged.

File: packages/go_router_builder/example/lib/shell_route_with_keys_example.dart (hunk 13 above)
Comment (GP4cK) [reply packages/go_router_builder/example/lib/shell_route_with_keys_example.dart:1]: I've added a simple test here: [Test shell_route_with_keys_example](https://github.com/flutter/packages/pull/3439/commits/28e593f917696da42a0e0169b247796a464f6b30)

File: packages/go_router_builder/example/pubspec.yaml (hunk 11 above)
Comment (chunhtai) [reply packages/go_router_builder/example/pubspec.yaml:15]: I see, this is fine then

**Review by chunhtai — APPROVED** [review]
LGTM

File: packages/go_router_builder/lib/src/route_config.dart [hunk 15]
```diff
@@ -105,6 +123,42 @@ class RouteConfig {
...
   final InterfaceElement _routeDataClass;
   final RouteConfig? _parent;
+  final String? _key;
+  final bool _isShellRoute;
+
+  static String? _generateNavigatorKeyGetterCode(
+    InterfaceElement classElement, {
+    required String keyName,
+  }) {
+    bool whereStatic(FieldElement element) => element.isStatic;
```
Comment (kevmoo) [inline packages/go_router_builder/lib/src/route_config.dart:133]: why create one-off functions for the tiny ones – but keep the big function inline?

Comment (GP4cK) [reply packages/go_router_builder/lib/src/route_config.dart:133]: No strong reason I believe.
Do you want me to make a PR to make everything inline?
Do I need to create an issue for this?

Comment (kevmoo) [reply packages/go_router_builder/lib/src/route_config.dart:133]: No issue needed. just a cleanup PR. I'd put everything inline

Comment (GP4cK) [reply packages/go_router_builder/lib/src/route_config.dart:133]: I grouped everything under one `.where` https://github.com/flutter/packages/pull/3698
